ServerClusterInterface: decouple Diagnostic Logs cluster.

pimpalemahesh opened the pull request:
#### Summary
Creates a DefaultServerCluster out of the DiagnosticLogsCluster and adds some unit tests as well.
- migrates src/app/cluster/diagnostic-logs-server to be a code driven cluster
- adds some basic unit tests

#### Related issues
Fixes #40495

#### Testing
- Unit test exists
- CI is expected to pass

pimpalemahesh committed:
Add tests for diagnostic logs cluster

andy31415 commented on examples/all-clusters-app/linux/main-common.cpp:
```diff
@@ -255,17 +262,6 @@ void emberAfWindowCoveringClusterInitCallback(chip::EndpointId endpoint)
 }
 
-using namespace chip::app::Clusters::DiagnosticLogs;
-void emberAfDiagnosticLogsClusterInitCallback(chip::EndpointId endpoint)
```
I would keep this integration. If we switch CodegenIntegration to use the `Server` integration method, this call will still work.

We should try to minimize API disruptions as much as possible. I would even consider making diagnostic-logs-server.h a shim that includes CodegenIntegration.

pimpalemahesh replied:
@andy31415 Currently, all examples set the delegate in emberAfDiagnosticLogsClusterInitCallback(). However, this callback is invoked before emberAfDiagnosticLogsClusterServerInitCallback(), meaning the cluster object has not been constructed yet. As a result, the delegate assignment occurs too early and gets lost, leading to a nullptr when the RetrieveLogsRequest command is invoked.

I think we should move the delegate setup to the application init instead.

andy31415 commented on examples/light-switch-app/qpg/src/AppTask.cpp:
```diff
@@ -93,6 +94,9 @@ void AppTask::InitServer(intptr_t arg)
     }
+
+    auto & logProvider = LogProvider::GetInstance();
+    chip::app::Clusters::DiagnosticLogs::SetDelegate(kRootEndpointId, &logProvider);
```
same here: we should aim at minimal API disruption.

It may also happen that we cannot preserve the API, in which case we should update https://github.com/project-chip/connectedhomeip/blob/master/docs/upgrading.md

pimpalemahesh committed:
Codegen Integration changes
- Introduce shim diagnostic logs server file
- declare dynamic array for delegate table

pimpalemahesh committed:
Initial implementation of code driven diagnostic logs cluster

pimpalemahesh committed:
Restyled changes. fix ci and typos

pimpalemahesh committed:
Update codegenIntegration as per the Centralize codegen integration approach

pimpalemahesh committed:
Run zap generation for current cluster changes

andy31415 commented on examples/light-switch-app/light-switch-common/light-switch-app.matter:
```diff
@@ -3107,7 +3107,7 @@ endpoint 0 {
   server cluster DiagnosticLogs {
     ram      attribute featureMap default = 0;
-    ram      attribute clusterRevision default = 1;
+    callback attribute clusterRevision;
```
featureMap should also be callback I believe - we have no feature maps to carry over from ember

pimpalemahesh committed:
Create singletone diagnostic logs cluster
    - use cluster server and shutdown callbacks

pimpalemahesh committed:
fix: Return command status as std::nullopt if status added in response

pimpalemahesh committed:
fix CI failure because of two Invalid command responses
- convert featuremap attribute to callback

andy31415 commented on src/app/clusters/diagnostic-logs-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,120 @@
+{
+    VerifyOrReturn(endpoint == kRootEndpointId,
+                   ChipLogError(DeviceLayer, "DiagnosticLogsProviderDelegate can only be set on root endpoint"));
+    VerifyOrReturn(delegate != nullptr, ChipLogError(DeviceLayer, "DiagnosticLogsProviderDelegate cannot be nullptr"));
```
why? I would drop the log and explain in a comment.

Does the delegate not start off as null? So why if I set it at start I cannot reset it back?

pimpalemahesh committed:
Addressed review comments

pimpalemahesh committed:
Restyled by whitespace

pimpalemahesh committed:
Fix tests failure

andy31415 committed:
Changes that likely decrease size:

- use a single global  singleton for CodegenIntegration. have less
  code/logic.
- Force the cluster to use kRootEndpoint

andy31415 committed:
Mark diagnostic logs revision and feature map as AAI.

andy31415 committed:
Some code clarity cleanup which saves some flash.

andy31415 committed:
Restyle

andy31415 commented on src/app/clusters/diagnostic-logs-server/Upgrading.md:
````diff
@@ -0,0 +1,41 @@
+New API:
+
+```
+chip::app::Clusters::DiagnosticLogsCluster::Instance().SetDelegate(delegate);
````
There is no Instance() call though.

If we use CodegenIntegration we may want to provide an instance. otherwise it is up to the app.

andy31415 commented on src/app/clusters/diagnostic-logs-server/Upgrading.md:
````diff
@@ -0,0 +1,41 @@
+Old API (deprecated):
+
+```
+chip::app::Clusters::DiagnosticLogs::DiagnosticLogsServer::SetDiagnosticLogsProviderDelegate(endpointId, delegate);
````
This seems untrue: the old and new code require another `Instance()` in the call. Should we correct it?

pimpalemahesh committed:
Add instance method, update readme

pimpalemahesh committed:
Remove global instance for classes

pimpalemahesh committed:
Documentation changes

pimpalemahesh committed:
Readme update

pimpalemahesh committed:
Merge logic into cluster class

pimpalemahesh committed:
Changes as per the migration doc

pimpalemahesh committed:
Change cluster server and shutdown callback

pimpalemahesh committed:
Add remaining zap file changes
- run zap_convert_all.py and zap_regen_all.py

pimpalemahesh committed:
fix tests: assert before optional value

pimpalemahesh committed:
fix ci

pimpalemahesh committed:
Added diagnostic cluster to code driven clusters in config-data.yaml
run zap regen
